velox/type/tests/TimestampTest.cpp

```diff
@@ -299,37 +299,129 @@ TEST(TimestampTest, outOfRange) {
+  ASSERT_EQ(expected.tm_min, actual.tm_min);
+  ASSERT_EQ(expected.tm_sec, actual.tm_sec);
+}
+
+std::string tmToString(
+    const std::tm& tmValue,
+    uint64_t nanos,
+    const std::string& format,
+    const TimestampToStringOptions& options) {
+  auto width = static_cast<int>(options.precision);
+  auto value =
+      options.precision == TimestampToStringOptions::Precision::kMilliseconds
+      ? nanos / 1'000'000
+      : nanos;
+
+  std::ostringstream oss;
+  oss << std::put_time(&tmValue, format.c_str());
+
+  if (options.mode != TimestampToStringOptions::Mode::kDateOnly) {
+    oss << '.' << std::setfill('0') << std::setw(width) << value;
+  }
+
+  return oss.str();
+}
+
 TEST(TimestampTest, epochToUtc) {
-  std::tm tm;
-  ASSERT_FALSE(epochToUtc(-(1ll << 60), tm));
-  ASSERT_FALSE(epochToUtc(1ll << 60, tm));
+  std::tm tm{};
+  ASSERT_FALSE(Timestamp::epochToUtc(-(1ll << 60), tm));
+  ASSERT_FALSE(Timestamp::epochToUtc(1ll << 60, tm));
 }
 
 TEST(TimestampTest, randomEpochToUtc) {
   uint64_t seed = 42;
-  // seed = std::random_device{}();
   std::default_random_engine gen(seed);
   std::uniform_int_distribution<time_t> dist(
       std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
-  std::tm actual, expected;
-  for (int i = 0; i < 10000; ++i) {
+  std::tm actual{};
+  std::tm expected{};
+  for (int i = 0; i < 10'000; ++i) {
+    auto epoch = dist(gen);
+    SCOPED_TRACE(fmt::format("epoch={}", epoch));
+    if (gmtime_r(&epoch, &expected)) {
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+    } else {
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
+    }
+  }
+}
+
+void testTmToString(
+    const std::string& format,
+    const TimestampToStringOptions& options) {
+  uint64_t seed = 42;
+  std::default_random_engine gen(seed);
+  std::default_random_engine nanosGen(seed);
+
+  std::uniform_int_distribution<time_t> dist(
+      std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
+  std::uniform_int_distribution<int> nanosDist(0, Timestamp::kMaxNanos);
+
+  std::tm actual{};
+  std::tm expected{};
+
+  for (int i = 0; i < 10'000; ++i) {
     auto epoch = dist(gen);
+    auto nanos = nanosDist(nanosGen);
     SCOPED_TRACE(fmt::format("epoch={}", epoch));
     if (gmtime_r(&epoch, &expected)) {
-      ASSERT_TRUE(epochToUtc(epoch, actual));
-      ASSERT_EQ(expected.tm_year, actual.tm_year);
-      ASSERT_EQ(expected.tm_yday, actual.tm_yday);
-      ASSERT_EQ(expected.tm_mon, actual.tm_mon);
-      ASSERT_EQ(expected.tm_mday, actual.tm_mday);
-      ASSERT_EQ(expected.tm_wday, actual.tm_wday);
-      ASSERT_EQ(expected.tm_hour, actual.tm_hour);
-      ASSERT_EQ(expected.tm_min, actual.tm_min);
-      ASSERT_EQ(expected.tm_sec, actual.tm_sec);
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+
+      auto actualMillisStr = Timestamp::tmToString(actual, nanos, options);
```

Comment:
> actualMillisStr

Is it accurate to use "Millis" in the variable name given different values of 'format' ? Should we rename it to actualString and expectedString?

Reply:
You are right, it is not appropriate to use `Millis` in variable names, I will fix it.